### DISCUSSION TITLE
[Question]: Docker MongoDB container "Operation not permitted" error on WSL 2 Windows 11

### DISCUSSION THREAD

**Discussion by lam584** [danny-avila/LibreChat#2053] [Troubleshooting] [answered]
### What is your question?

I'm encountering a persistent "Operation not permitted" error when attempting to start a MongoDB container with Docker on Windows 11 using WSL 2. The MongoDB container is set to use a mounted volume from the Windows filesystem (`/mnt/d/LibreChat/data-node`). Despite the directory and its files showing full read-write-execute (`rwxrwxrwx`) permissions when checked from the WSL terminal, MongoDB fails to start, citing permission issues.

## Environment
- OS: Windows 11 22631.3235
- Docker Desktop Version: 4.27.2 (137060)
- WSL Version: 2
- MongoDB Docker Image Version: Latest (as of date of issue posting)
- Issue encountered on: `/mnt/d/LibreChat/data-node`

## Description
I'm encountering a persistent "Operation not permitted" error when attempting to start a MongoDB container with Docker on Windows 11 using WSL 2. The MongoDB container is set to use a mounted volume from the Windows filesystem (`/mnt/d/LibreChat/data-node`). Despite the directory and its files showing full read-write-execute (`rwxrwxrwx`) permissions when checked from the WSL terminal, MongoDB fails to start, citing permission issues.

## Steps to Reproduce
1. Navigate to the directory containing the `docker-compose.yml`.
2. Run `docker compose up`.
3. MongoDB container fails to start with an "Operation not permitted" error.

## Expected Behavior
MongoDB container starts successfully, accessing and manipulating data within the specified directory without permission issues.

## Actual Behavior
MongoDB fails to start, with log output suggesting permission-related errors accessing its data directory mounted from the Windows file system. 

## Additional Context
- I have verified that the directory and files at `/mnt/d/LibreChat/data-node` have `rwxrwxrwx` permissions (using `ls -l`), and the owner is `root`.
- Here is a snippet of the MongoDB related section from my `docker-compose.yml`:
    ```yaml
    mongodb:
      container_name: chat-mongodb
      image: mongo
      restart: always
      volumes:
        - ./data-node:/data/db
      command: mongod --noauth
    ```
- I have checked Docker and WSL settings regarding file sharing and permissions but found no clues pointing to the cause of this issue.

Given these points, it seems the problem might be related to how Docker/WSL 2 handles permissions for Windows filesystem-mounted volumes, or a specific interaction with the MongoDB container. Any assistance or insights from the community would be greatly appreciated!

Thank you!

### What is the main subject of your question?

Installation

### Screenshots

![20240310211408](https://github.com/danny-avila/LibreChat/assets/107453273/13db38e3-ea1f-456e-a274-5c2961450788)
![20240310211409](https://github.com/danny-avila/LibreChat/assets/107453273/20c0e0e2-bcd5-41ef-9def-eedbe9d1bfa6)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by james-williams0** [comment]
Also observing a similar issues with a similar environment, using the docker-compose.yml as included with the LibreChat repo and its own `chat-mongodb` container.

Environment:

- Windows 11 23H2 Build 22631.3235
- Docker Desktop 4.28.0 (139021)
- WSL 2
- Commit hash `0f417aae`

Pulled the following logs out which seem relevant:
```
chat-mongodb      | {"t":{"$date":"2024-03-11T23:12:35.774+00:00"},"s":"I",  "c":"WT",       "id":22430,   "ctx":"initandlisten","msg":"WiredTiger message","attr":{"message":{"ts_sec":1710198755,"ts_usec":773993,"thread":"1:0x7fb38f50dc80","session_name":"connection","category":"WT_VERB_BLOCK","category_id":3,"verbose_level":"NOTICE","verbose_level_id":-1,"msg":"unexpected file WiredTiger.wt found, renamed to WiredTiger.wt.17"}}}
chat-mongodb      | {"t":{"$date":"2024-03-11T23:12:35.774+00:00"},"s":"E",  "c":"WT",       "id":22435,   "ctx":"initandlisten","msg":"WiredTiger error message","attr":{"error":1,"message":{"ts_sec":1710198755,"ts_usec":774925,"thread":"1:0x7fb38f50dc80","session_name":"connection","category":"WT_VERB_DEFAULT","category_id":9,"verbose_level":"ERROR","verbose_level_id":-3,"msg":"__posix_open_file:815:/data/db/WiredTiger.wt: handle-open: open","error_str":"Operation not permitted","error_code":1}}}
chat-mongodb      | {"t":{"$date":"2024-03-11T23:12:35.776+00:00"},"s":"W",  "c":"STORAGE",  "id":22347,   "ctx":"initandlisten","msg":"Failed to start up WiredTiger under any compatibility version. This may be due to an unsupported upgrade or downgrade."}
chat-mongodb      | {"t":{"$date":"2024-03-11T23:12:35.776+00:00"},"s":"F",  "c":"STORAGE",  "id":28595,   "ctx":"initandlisten","msg":"Terminating.","attr":{"reason":"1: Operation not permitted"}}
chat-mongodb      | {"t":{"$date":"2024-03-11T23:12:35.776+00:00"},"s":"F",  "c":"ASSERT",   "id":23091,   "ctx":"initandlisten","msg":"Fatal assertion","attr":{"msgid":28595,"file":"src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp","line":676}}
chat-mongodb      | {"t":{"$date":"2024-03-11T23:12:35.776+00:00"},"s":"F",  "c":"ASSERT",   "id":23092,   "ctx":"initandlisten","msg":"\n\n***aborting after fassert() failure\n\n"}
chat-mongodb exited with code 14
```

**Reply by fuegovic** [reply]
Try uncommenting UID and GID in the .env file

**Reply by james-williams0** [reply]
It's uncommented, set to the default values of `1000`:
```
#=============#
# Permissions #
#=============#

UID=1000
GID=1000
```

**Reply by james-williams0** [reply]
For completeness, I've gone back and tried it with it commented, same result.

**Reply by fuegovic** [reply]
~~if they were uncommented, then try commenting them out instead~~

**Reply by fuegovic** [reply]
ok... one or the other often fix WiredTiger issue, tbh I'm not sure what's the solution in your case then

**Reply by lam584** [reply]
I have tried to uncomment UID and GID, but it cannot solve the problem.
![20240310211408](https://github.com/danny-avila/LibreChat/assets/107453273/85592e67-0b2c-4b7a-858c-22b4fd2fb3ce)
![20240310211409](https://github.com/danny-avila/LibreChat/assets/107453273/f0f18f2c-f30b-4712-893b-e5ee8ba8b66e)

**Reply by james-williams0** [reply]
I have successfully worked around my issue by cloning and running it under an [Ubuntu 22.04 install on WSL](https://apps.microsoft.com/detail/9pdxgncfsczv?rtc=1&hl=en-gb&gl=GB), _with the UID and GID commented in the_ `.env` _file_, using the [Linux installation guide](https://docs.librechat.ai/install/installation/linux_install.html).

**Comment by halllo** [comment, marked as answer]
I am running podman on Windows and had the same MongoDB error:

```logs
"category_id":12,"verbose_level":"ERROR","verbose_level_id":-3,
"msg":"__posix_open_file:924:/data/db/WiredTiger.wt: handle-open: open",
"error_str":"Operation not permitted","error_code":1
```

Its reproducible independent of LibreChat:

```powershell
podman run -d -v ./dn:/data/db mongo mongod --noauth
```

This summary about [podman-mongodb](https://github.com/eriksjolund/podman-mongodb) helped me understand the situation better. So I was able to make mongodb work with `--userns keep-id`:

```powershell
podman run -d --name self_mongodb --userns keep-id -v ./dn:/data/db mongo mongod --noauth
```

I added this to the `docker-compose.override.yml`:

```yml
services:
  mongodb:
    userns_mode: "keep-id"
```

Unfortunately now LibreChat could not be composed anymore:

```logs
234e06dbdaca3f5e73be8b8439996c8d19670cfcd1830b6689ddeebbed89d1d2
Error: --userns and --pod cannot be set together
...
```

But then I found https://github.com/containers/podman-compose/issues/166 and `--in-pod`. So with the following command I was finally able to start up LibreChat with podman rootless containers using podman-compose on Windows:

```powershell
uvx podman-compose --in-pod false up -d
```
